*: add CollectProfile method to the Resumer

adityamaru opened the pull request:
Please refer to individual commits.

Closes: https://github.com/cockroachdb/cockroach/issues/109671

adityamaru committed:
*: add CollectProfile method to the Resumer

This change adds a new method to the Resumer interface
that will be called whenever we request the execution details
for a particular job. This method is currently stubbed for
all resumers but will be filled out for those jobs that have
opted in to jobs profiling.

Informs: #109671
Release note: None

adityamaru committed:
backupccl: tracing aggregator stats are now dumped in CollectProfile

Previously, tracing aggregator stats sent back by all processors in
the flow would be periodically dumped to the job_info table basis an
arbitrary timer. Now, these stats are only persisted to disk when the
job has been instructed to collect a profile.

Informs: #109671
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/110556)
<!-- Reviewable:end -->

miretskiy commented on pkg/jobs/registry.go:
```diff
@@ -1454,6 +1454,11 @@ type Resumer interface {
+	// CollectProfile is called when a job has been requested to collect a job
+	// profile. This profile may contain any information that provides enhanced
+	// observability into a job's execution.
+	CollectProfile(ctx context.Context, execCtx interface{}) error
```
Do we want to have every resumer implement this?
Looks like most of them return nil.
Perhaps we can have 
```
type ProfilerCollector interface {
   CollectProfile(....)
}
```
And have the jobs implement this interface?
When you look at the resumer, check to see if it implemented this method, and call it.  If not, return "nil" as the default implementation already does.

dt replied:
I think I prefer the implementation actively, explicitly state that it doesn't want to do anything when profiled, rather than implicitly say so by forgetting to implement the method.

adityamaru committed:
backupccl: restore aggregator stats are now dumped in CollectProfile

Previously, tracing aggregator stats sent back by all processors in
the flow would be periodically dumped to the job_info table basis an
arbitrary timer. Now, these stats are only persisted to disk when the
job has been instructed to collect a profile.

Informs: #109671
Release note: None

adityamaru committed:
streamingccl: implement CollectProfile

Previously, tracing aggregator stats sent back by all processors in
the flow would be periodically dumped to the job_info table basis an
arbitrary timer. Now, these stats are only persisted to disk when the
job has been instructed to collect a profile.

Informs: #109671
Release note: None

adityamaru committed:
bulk: delete timer based aggregator stats helper

dt commented on pkg/server/job_profiler.go:
```diff
@@ -78,8 +79,20 @@ func (s *statusServer) RequestJobProfilerExecutionDetails(
-		// specific execution details collection logic.
+		r, err := execCfg.JobRegistry.GetResumerForClaimedJob(jobID)
+		if err != nil {
+			// Request job execution details is best effort, so swallow the error
```
Huh, why not report this back to the user?

adityamaru replied:
A few tests were failing because the resumers weren't in the registry's map, so I convinced myself its okay to present as much information as we can. But I realize this means that we'd never see errors this way. Let me think some more and see if the tests are easier to fix.

adityamaru replied:
done

dt commented on pkg/ccl/backupccl/backup_job.go:
```diff
@@ -1230,8 +1248,16 @@ func (b *backupResumer) OnFailOrCancel(
+
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	return bulkutil.FlushTracingAggregatorStats(ctx, b.job.ID(),
```
This is potentially a going to hold the `mu` for some time. Could consider copying the map -- both levels -- to release the mu before passing the copy to flush, since the mu blocks draining the progress ch which can backpressure the flow can't it?

adityamaru replied:
Oh interesting so if we don't flush tracingAggCh fast enough, we could backup the processors from sending progress metas back to us. I didn't think of that. I'm cool with copying the map and passing that in to flush.

adityamaru commented:
TFTR!

bors r=dt
